### ISSUE TITLE
FR: implement DNS64 for 4via6 subnet routing

### ISSUE THREAD

**Issue by MrGossett** [tailscale/tailscale#5392]
### What are you trying to do?

I'm using Tailscale as an identity-aware proxy to provide access to a non-public environment in AWS. I have a subnet router deployed in the VPC, advertising routes to the VPC's CIDR (`10.0.0.0/16`). I also have my Tailnet configured with split DNS to send queries for the environment's domain (`example.com`) to the VPC's resolver (`10.0.0.2`). This is working as advertised. Machines on my Tailnet send a DNS query through the subnet router to the VPC's resolver, which resolves `example.com` from the alias A record for an internal load balancer; a set of IPv4 addresses in the `10.0.0.0/16` range come back in the response, and the HTTP request that follows is properly routed through the subnet router and into the VPC.

I would like to expand this setup to provide access to another non-public environment in a different VPC. Both VPCs are using the same CIDR (`10.0.0.0/16`), so this seems like a use case for Tailscale's 4via6 routing. I'm excited about the 4via6 feature, and I'm interested to try it out while it's in alpha/beta.

I can have the subnet routers in both VPCs advertise 4via6 routes:
* the subnet router in the first VPC would advertise a route to `fd7a:115c:a1e0:b1a:0:1:a00:0/112` for site 1
* the subnet router in the second VPC would advertise a route to `fd7a:115c:a1e0:b1a:0:2:a00:0/112` for site 2

I would update my Tailnet's DNS configuration like this:
* queries for `example.com` are sent to `fd7a:115c:a1e0:b1a:0:1:a00:2`, which maps to `10.0.0.2` in site 1
* queries for `example.net` are sent to `fd7a:115c:a1e0:b1a:0:2:a00:2`, which maps to `10.0.0.2` in site 2

The problem is that the VPC resolvers respond with IPv4 addresses. There's nothing in this setup that would translate IPv4 addresses in an A-record response into their corresponding 4via6 IPv6 addresses in a AAAA-record response. So with this setup, users in my Tailnet would be able to resolve the private IPv4 addresses of the load balancers in both sites, but since those addresses are not mapped to their 4via6 counterparts, the resulting traffic would not be properly routed.

Clearly, I need to add something.

### How should we solve this?

I'm not entirely sure how this should work, but I think I'm asking for [DNS64](https://datatracker.ietf.org/doc/html/rfc6147) functionality in Tailscale's resolver. Tailscale is already handling some other DNS things, so this functionality seems like it fits.

### What is the impact of not solving this?

The 4via6 feature solves a problem at the IP level, but the problem I've described here is at the DNS level. Without this feature, 4via6 would only be useful for IP-only use cases, or for users that are okay with manually maintaining 4via6 addresses in a DNS zone.

I think implementing this feature would enable a complete identity-aware proxy solution for access to protected environments, powered by Tailscale.

### Anything else?

The [DNS64](https://datatracker.ietf.org/doc/html/rfc6147) spec seems to define the behavior I'm looking for in this issue. Clients in my Tailnet are constrained to be IPv6-only because they need to deal exclusively in 4via6 address space, and servers in my VPC are IPv4-only. The spec [describes](https://datatracker.ietf.org/doc/html/rfc6147#page-7) "DNS64 in stub-resolver mode" where translation happens on the initiator endpoint; that sounds like something happening in Tailscale's resolver. If I were to roll my own solution, it would likely be what the spec calls "DNS64 in DNS recursive-resolver mode," adding a recursive resolver somewhere between my clients and servers that would do the 4via6 mapping.

[AWS supports DNS64](https://docs.aws.amazon.com/vpc/latest/userguide/nat-gateway-nat64-dns64.html), but that's for a scenario where IPv6-only clients _inside_ the VPC are talking to IPv4-only clients _outside_ the VPC. I'm trying to do the inverse: IPv6 clients _outside_ the VPC (i.e., in my Tailnet) talking to IPv4-only clients _inside_ the VPC.

I had considered writing a small DNS resolver that would use [`tsaddr.MapVia`](https://pkg.go.dev/tailscale.com@v1.28.0/net/tsaddr#MapVia) to translate IPv4 addresses in a DNS response to their IPv4 counterparts. It would be configured with a site ID and IPv4 prefixes indicating which addresses should be mapped (so only addresses within `10.0.0.0/16` would be translated in my example), and everything else would be sent to an upstream resolver.

Then I learned (thanks, [O'Reilly](https://www.oreilly.com/library/view/dns-and-bind/9781449308025/ch04.html)!) that `bind` supports DNS64. So perhaps I could run a `bind` server with config like the following, and have my Tailnet split DNS traffic to `bind` for `example.com` and `example.net`. This is presently the strongest option I'm aware of (and perhaps a good addition to Tailscale docs on 4via6), but it means I need to run `bind` somewhere.

Here's a `bind` config snippet that I think would work for my example:

```bind
    # site 1
    dns64 fd7a:115c:a1e0:b1a:0:1::/96 {
        mapped { 10.0/16; };
    };

    # site 2
    dns64 fd7a:115c:a1e0:b1a:0:2::/96 {
        mapped { 10.0/16; };
    };
```

I would have to use this config on a single `bind` server in my Tailnet, outside of the target VPCs, or I would run a `bind` server in each VPC with config only relevant to that VPC. Either way, I would need to step outside of Tailscale to complete the solution.

**Comment by MrGossett** [comment]
related: https://github.com/tailscale/terraform-provider-tailscale/pull/136

**Comment by MrGossett** [comment]
Update: I have completed a proof-of-concept for the option that involves running a bind server, and it's working for the use case I described above. More details are below the fold. If Tailscale decides to implement the functionality I'm asking for here, perhaps the sharp corners I ran into can inform the implementation.

TL;DR: while the bind option is effective, I didn't find this to be a user-friendly option.

<details>
<summary>Bind config</summary>

I needed to accommodate the behavior of clients that will first ask for an A record, and then follow up with a request for a AAAA record if there was no A record found. My first pass at using bind with its `dns64` option didn't work: it would synthesize AAAA answers correctly if the question was asking for AAAA records; but it would also happily return an answer with private-range IPv4 addresses in response to a question for A records. I could use the `deny-answer-addresses` option to tell bind not to include IPv4 addresses in answers, and I could use the `dns64` option to synthesize IPv6 answers for AAAA questions, but I couldn't use both options at the same time. The `deny-answer-addresses` option would take effect before the `dns64` option had synthesized its answer.

I ended up configuring bind to listen on two ports and forwarding questions to itself before forwarding on to the VPC resolver. This way, I could use the `deny-answer-addresses` option on one listener, and use the `dns64` option on the other listener. Effectively, this makes the `deny-answer-addresses` option apply to answers after the `dns64` option has synthesized an answer.

Some context: in this POC, I had
* VPC with CIDR `10.0.0.0/16`
* 4via6 subnet router running in the VPC advertising `fd7a:115c:a1e0:b1a:0:2:a00:0/112` (site `2`, CIDR `10.0.0.0/16`)
* `bind` running on another EC2 (`10.0.2.60`) in the same VPC
* my Tailnet splitting DNS to the `bind` server's 4via6 IP (`fd7a:115c:a1e0:b1a:0:2:a00:23c`)
* my local machine as another node in the Tailnet

```conf
# /etc/bind/named.conf

acl outside { !127.0.0.1; any; }; # anything that's not localhost is on the outside
acl inside  {  127.0.0.1; };      # localhost is on the inside

options {
	listen-on port 53   { outside; }; # listen on port 53 for outside interfaces
	listen-on port 5353 { inside; };  # listen on port 5353 on the loopback interface

	directory "/var/cache/bind"; # bind needs a place to write cache stuff; the default is "."
	dnssec-validation yes;       # without this, bind complains about a "broken trust chain"
};

view "outside" {
	match-clients         { outside; };   # this view applies to clients coming from the outside
	deny-answer-addresses { 10.0/16; };   # it denies any answers that have IPv4 addresses
	forwarders port 5353  { 127.0.0.1; }; # questions are forwarded to localhost on port 5353
	forward only;                         # don't try to be an authority
	recursion yes;                        # allow recursive queries
};

view "inside" {
	match-clients { inside; };   # this view applies to clients coming from the inside
	forwarders    { 10.0.0.2; }; # questions are forwarded to the VPC resolver at the +2 address
	forward only;                # don't try to be an authority
	recursion yes;               # allow recursive queries

	# DNS64 will synthesize AAAA answers from A records
	dns64 fd7a:115c:a1e0:b1a:0:2::/96 {
		mapped { 10.0/16; };  # only synthesize answers for addresses in this range
		break-dnssec yes;     # don't worry about DNSSEC
	};
};
```

From my local machine, if I query the VPC resolver at `fd7a:115c:a1e0:b1a:0:2:a00:2` (which is the 4via6 address for `10.0.0.2`), I get back private-range IPv4 addresses of a load balancer:

```
$ dig @fd7a:115c:a1e0:b1a:0:2:a00:2 A app.example.com

; <<>> DiG 9.10.6 <<>> @fd7a:115c:a1e0:b1a:0:2:a00:2 A app.example.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 5055
;; flags: qr rd ra; QUERY: 1, ANSWER: 3, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;app.example.com.	IN	A

;; ANSWER SECTION:
app.example.com. 41	IN	A	10.0.3.134
app.example.com. 41	IN	A	10.0.14.255
app.example.com. 41	IN	A	10.0.23.196

;; Query time: 27 msec
;; SERVER: fd7a:115c:a1e0:b1a:0:2:a00:2#53(fd7a:115c:a1e0:b1a:0:2:a00:2)
;; WHEN: Wed Aug 17 20:13:21 EDT 2022
;; MSG SIZE  rcvd: 100
```

When I ask the bind server (via `fd7a:115c:a1e0:b1a:0:2:a00:23c`) for an A record for that same hostname, I get no response:

```
$ dig @fd7a:115c:a1e0:b1a:0:2:a00:23c A app.example.com

; <<>> DiG 9.10.6 <<>> @fd7a:115c:a1e0:b1a:0:2:a00:23c A app.example.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 7543
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;app.example.com.       IN      A

;; Query time: 28 msec
;; SERVER: fd7a:115c:a1e0:b1a:0:2:a00:23c#53(fd7a:115c:a1e0:b1a:0:2:a00:23c)
;; WHEN: Wed Aug 17 20:08:10 EDT 2022
;; MSG SIZE  rcvd: 52
```

When I ask bind for a AAAA record, I get the DNS64 mapped response:

```
$ dig @fd7a:115c:a1e0:b1a:0:2:a00:23c AAAA app.example.com
; <<>> DiG 9.10.6 <<>> @fd7a:115c:a1e0:b1a:0:2:a00:23c AAAA app.example.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 56915
;; flags: qr rd ra; QUERY: 1, ANSWER: 3, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;app.example.com.       IN      AAAA

;; ANSWER SECTION:
app.example.com. 60     IN      AAAA    fd7a:115c:a1e0:b1a:0:2:a00:386
app.example.com. 60     IN      AAAA    fd7a:115c:a1e0:b1a:0:2:a00:17c4
app.example.com. 60     IN      AAAA    fd7a:115c:a1e0:b1a:0:2:a00:eff

;; Query time: 62 msec
;; SERVER: fd7a:115c:a1e0:b1a:0:2:a00:23c#53(fd7a:115c:a1e0:b1a:0:2:a00:23c)
;; WHEN: Wed Aug 17 20:08:02 EDT 2022
;; MSG SIZE  rcvd: 136
```

And those IPv6 addresses match exactly to the private-range IPv4 addresses that the VPC resolver returned.

```
$ dig +short @fd7a:115c:a1e0:b1a:0:2:a00:23c AAAA app.example.com | while read ipv6; do tailscale debug via "$ipv6/128"; done
site 2 (0x2), 10.0.23.196/32
site 2 (0x2), 10.0.14.255/32
site 2 (0x2), 10.0.3.134/32
```

</details>

**Comment by maisem** [comment]
Tailscale already has MagicDNS support for this. Does that not cover your use case?
https://tailscale.com/kb/1201/4via6-subnets/#magicdns-name-for-the-ipv4-subnet-devices

**Comment by MrGossett** [comment]
@maisem it does not. The MagicDNS support for 4via6 routing supports hostnames like `Q.R.S.T.via-X`. In my use case, the IPV4 address `Q.R.S.T` would be one of the addresses of a load balancer. In order to support the end-to-end behavior I'm after, I would need to CNAME `example.com` to `Q.R.S.T.via-X`, pinning the load balancer to just one of its IP addresses; and I would need to make sure the CNAME record stays up to date as the load balancer's IPv4 address changes.

**Comment by PercussiveElbow** [comment]
Running into this same issue - it would be ideal if this would be something that could be specified in the Split DNS settings in Magic DNS, so that queries for say internal.com would route to an internal DNS server over 4via6 subnet routing which would return IPv6 results that 4via6 could handle rather than whatever IPv4 records it might otherwise return.

i.e.

With Magic DNS turned on and 4v6 routing enabled to 10.0.0/24 at site # 1:
- Within the admin panel specify that SplitDNS is configured to route _internal.com_ to a DNS server at 10.0.0.1 at a site # 1
- A client on the Tailnet queries for example.internal.com, which queries the DNS  server at 10.0.0.1.via-1 which would ordinarily return a record corresponding to an an IPv4 address on that network e.g. 10.0.0.2.
- Ideally what could then happen is MagicDNS modifies this response to then return a 4via6 format (10.0.0.2.via-1) which is then resolved to the correct IP address, assuming the IPv4 result was within the subnet router range and not something external.


Without this I think there's no choice but to start hard coding DNS records or running another DNS server in the mix like suggested above if I'm not missing something - I think this takes away some of the magic of 4via6 which is a pretty incredible feature otherwise. There may be a more sensible solution than what I've suggested though.

**Comment by DentonGentry** [comment]
It seems like this could be implemented in a standalone DNS server outside of tailscaled as well, though it would need to coordinate its lookups with MagicDNS to get the via-N addresses.

**Comment by b-** [comment]
I implemented something like this in [a custom CoreDNS plugin](https://github.com/briorg/coredns-a2cname). It's "not" DNS64, because I wasn't aware of DNS64 until writing this, so I didn't know there was a spec I should be following! Aside from that, my scope was a little limited and MagicDNS does most of the translation for this anyway:

My plugin allows you to run a CoreDNS instance that internally translates a given (sub)domain's A records into synthetic CNAME records that themselves point to MagicDNS's `Q-R-S-T-via-X.my-tailnet.ts.net` AAAA records. I figured that receiving CNAME records might make more sense than receiving AAAA records in response to A requests (and frankly it was simpler to write).

The way I'm using it is that I have a subnet route for 10.0.0.0/16, which I wanted to eventually retire in place of a 4via6 route via site ID 9, but there are a number of internal DNS domains that point to internal IP addresses on that site's subnet. So, I wrote a DNS server plugin where I configure it to translate the A records into CNAME records. An example follows, where `some-vm.my-internal-domain.com` resolves to the IPv4 address `10.0.4.3`, and I use my plugin to remap it to MagicDNS's 4via6 site 9. My tailnet domain is `shark-perch.ts.net`:

```console
❯ nslookup some-vm.my-internal-domain.com
Server:         127.0.0.53
Address:        127.0.0.53#53

Non-authoritative answer:
some-vm.my-internal-domain.com    canonical name = 10-0-4-2-via-9.shark-perch.ts.net.
Name:   10-0-4-2-via-9.shark-perch.ts.net
Address: fd7a:115c:a1e0:b1a:0:9:a00:402
```

Of course the best way to solve overlapping IPv4's is to never have them in the first place :)
